File: templates/problems/11_graphs/09_course_schedule_ii/typescript/solution.ts
/**
 * Return a topological ordering of courses using Kahn's algorithm (BFS).
 * Time: O(V + E), Space: O(V + E).
 */
export function solveFindOrder(numCourses: number, prerequisites: number[][]): number[] {
    const graph: number[][] = Array.from({ length: numCourses }, () => []);
    const inDegree = new Array(numCourses).fill(0);
    for (const p of prerequisites) {
        graph[p[1]].push(p[0]);
        inDegree[p[0]]++;
    }

    const queue: number[] = [];
    for (let i = 0; i < numCourses; i++) {
        if (inDegree[i] === 0) queue.push(i);
    }
    const order: number[] = [];
    let head = 0;
    while (head < queue.length) {
        const node = queue[head++];
        order.push(node);
        for (const nxt of graph[node]) {
            inDegree[nxt]--;
            if (inDegree[nxt] === 0) queue.push(nxt);
        }
    }

    if (order.length !== numCourses) return [];
    return order;
}

File: templates/problems/11_graphs/09_course_schedule_ii/python/solution.py
from collections import deque


def solve_find_order(num_courses: int, prerequisites: list[list[int]]) -> list[int]:
    """Return a topological ordering of courses using Kahn's algorithm (BFS).

    Time: O(V + E), Space: O(V + E).
    """
    graph: list[list[int]] = [[] for _ in range(num_courses)]
    in_degree = [0] * num_courses
    for p in prerequisites:
        graph[p[1]].append(p[0])
        in_degree[p[0]] += 1

    queue: deque[int] = deque(i for i in range(num_courses) if in_degree[i] == 0)
    order: list[int] = []

    while queue:
        node = queue.popleft()
        order.append(node)
        for nxt in graph[node]:
            in_degree[nxt] -= 1
            if in_degree[nxt] == 0:
                queue.append(nxt)

    if len(order) != num_courses:
        return []
    return order
